### PR TITLE
Add Region file reading support

### PR DESCRIPTION
**Pull request by piegamesde**
This PR adds a new `RegionFile` class that makes it easy to access NBT data of chunks within a Minecraft Region/Anvil file. Also added some tests.

This pull requests makes use of the changes in #12 and the test won't work until #10 is merged since 1.13 NBT data is tested too.

### CODE REVIEW

**Comment by xcube16** [comment]
https://github.com/flow/nbt/blob/0dde724db4e363dd6db04d2491436167dbdc68e2/src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java

This is a state machine NIGHTMARE!

RegionFile#loadAllChunks() ick!
RegionFile#unloadAllChunks() ick!!
At first glance I thought RegionChunk#unload() and RegionFile#unloadAllChunks() saved the chunks,
but on closer inspection, someone is trying to re-invent the garbage collector!

The only methods should be:
RegionFile#getChunk(x, z) that reads and returns an Optional\<RegionChunk\>
maybe some kind of iterator functionality to iterate over the RegionChunks
and when writing chunks is supported (optional):
RegionFile#setChunk(x, z, RegionChunk)
RegionFile#clearChunk(x, z)

Making RegionChunk immutable-ish might help, but whatever works best.

Edit: And don't bother caching RegionChunk's at all. Let the user do that with a wrapper if they really need that. If someone is dumb enough to load the same chunk 1000 times the OS caches hot files anyway.

**Comment by piegamesde** [comment]
@xcube16 You're totally right. I changed those things a few months ago, but forgot to push them to the branch. Sadly, a few other changes have sneaked in and I don't really remember the details, I hope it's still ok.

But thanks for reading through my code in the first place :)

**Comment by xcube16** [comment]
@piegamesde Your welcome. :)
I came here to see if my own bug had been fixed by other PRs.

~~Now that I notice it, it looks like there is already a RegionFile reading implementation at https://github.com/flow/nbt/blob/develop/src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java
But it could use some work.~~

It looks like the last commit to the develop branch was 2 years ago. I hope a maintainer is still active on this project.

File: src/main/java/com/flowpowered/nbt/regionfile/Chunk.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import com.flowpowered.nbt.ListTag;
+import com.flowpowered.nbt.Tag;
+import com.flowpowered.nbt.stream.NBTInputStream;
+import com.flowpowered.nbt.stream.NBTOutputStream;
+
+/**
+ * Each instance of this class represents a Minecraft chunk within a Region file. The data is represented as a binary blob in form of a
+ * {@link ByteBuffer}. Each object is self-contained and not linked to any physical file or other object. Each object is immutable and
+ * should be treated as such. The data is set in the constructor</br>
+ * The data must have a length of a multiple of 4096 bytes (to be able to write it to disk more easily). The four first bytes specify the
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/Chunk.java:25]: The internal format of a region file should not be exposed to the user. Things like sector alignment, etc.

The blob has a format of
[length | 4 bytes, big endian][compression_flag | byte][data | length - 1 bytes]
'length' is data's length + 1, 'length' == 0 is a special case (a non existing chunk, don't ask my why mojang has this when normally the offset into the region file is 0 when there is a non existing chunk)
'compression_flag' == 1 means GZIP(Input/Output)Stream is used for compression (unused by minecraft to save chunks, but I would support it anyway),
'compression_flag' == 2 means Inflater(Input/Output)Stream is used for compression

A user will not realize this and try to use NBT(Input/Output)Stream on it directly and get a big surprize!
I would instead expose the uncompressed data (I know NBT(Input/Output)Stream supports compression with GZIP(Input/Output)Stream, but lets not force the user to deal with special cases)

File: src/main/java/com/flowpowered/nbt/regionfile/Chunk.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+	 *            The number of the 4096 byte sector where the chunk is located in the file. Don't forget that the first five bytes are used to
+	 *            store the size and compression of the chunk. The position of the first byte of NBT data is thus {@code start*4096 + 5}.
+	 * @param length
+	 *            The amount of 4096 byte sectors to load. It should be large enough to contain all NBT data in that chunk or it will be
+	 *            corrupted.
+	 * @throws IllegalArgumentException
+	 *             if the coordinates are out of bounds
+	 * @author piegames
+	 */
+	public Chunk(int x, int z, int timestamp, FileChannel raf, int start, int length) throws IOException {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/Chunk.java]: This constructor should not be public, but instead be package private. (or maybe not even exist, but that is up to you)

File: src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java [hunk 3]
```diff
@@ -77,7 +78,7 @@
                 raf.readFully(data);
                 ByteArrayInputStream in = new ByteArrayInputStream(data);
                 InflaterInputStream iis = new InflaterInputStream(in);
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java]: See. SimpleRegionFileReader does not force NBTInputStream to implement a compression method that only it (region files) ever use. :)

File: src/main/java/com/flowpowered/nbt/stream/NBTInputStream.java [hunk 4]
```diff
@@ -54,6 +56,22 @@
...
+	/**
+	 * Flag indicating that the given data will be compressed with the GZIP compression algorithm. This is the default compression method used to compress nbt files.
+	 * Chunks in Minecraft Region/Anvil files with compression method {@code 1} (see the respective format documentation) will use this compression method too, although this is not actively used anymore.
+	 */
+	public static final int GZIP_COMPRESSION = 1;
+	/**
+	 * Flag indicating that the given data will be compressed with the ZLIB compression algorithm. This is the default compression method used to compress the nbt
+	 * data of the chunks in Minecraft Region/Anvil files, but only if its compression method is {@code 2} (see the respective format documentation), which is default for all newer versions.
+	 */
+	public static final int ZLIB_COMPRESSION = 2;
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/stream/NBTInputStream.java]: You did all this work and added all this extra code just to support a compression method that only region files use? o.O
I would just decompress the data in your region file abstraction and feed uncompressed data to NBTInputStream

File: src/test/java/com/flowpowered/nbt/RegionFileTest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+	/**
+	 * Test reading the NBT data in a region file
+	 *
+	 * @throws URISyntaxException
+	 * @throws IOException
+	 */
+	@Test
+	public void testRead() throws IOException, URISyntaxException {
+		RegionFile file = new RegionFile(Paths.get(getClass().getResource("/r.1.3.mca").toURI()));
```
Comment (xcube16) [inline src/test/java/com/flowpowered/nbt/RegionFileTest.java]: Do we really need to include large binary test files in this project? There is a LOT of unknown data in a minecraft region file. This test does not check the integrity of any of that data, all it does is go "yay! It did not crash!".

The best way to test would be to use a custom region file (or, even better! generate one in the test!) with a known and verifiable NBT structure, empty chunks, the works! Than you can test the NBT when you read it and make shore thing is corrupted.

File: src/test/java/com/flowpowered/nbt/stream/EndianSwitchableStreamTest.java [hunk 6]
```diff
@@ -44,9 +44,11 @@ public void testWriteLEUnsignedShort() throws IOException {
         EndianSwitchableOutputStream output = new EndianSwitchableOutputStream(rawOutput, ByteOrder.LITTLE_ENDIAN);
         output.writeShort(unsigned);
         output.writeChar(testChar);
+        output.close();
```
Comment (xcube16) [inline src/test/java/com/flowpowered/nbt/stream/EndianSwitchableStreamTest.java:46]: // just a random comment, not important for this review
Kind of ironic that there are tests for LITTLE_ENDIAN short and char but not float and double. This would have saved me 3hrs of debugging. :'(
(see #16)

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+			for (int i = 0; i < 32 * 32; i++) {
+				if (!chunksUpToDate[i]) {
+					toSave.add(i);
+					continue;
+				}
+				if (chunks[i] == null)
+					continue;
+				int chunkPos = locations2.get(i) >>> 8;
+				int chunkLength = locations2.get(i) & 0xFF;
+				usedSectors.set(chunkPos, chunkPos + chunkLength);
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: usedSectors is being populated with old state. This will not cause corruption (I think), but it does mean that any mutated chunks will be likely pushed to the end of the file even though they could fit in there old location. (wasted space)

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+ * file will not be modified externally in any ways. All changes to the objects are kept in memory and only written back to the file upon
+ * calling {@link #writeChanges()}, in which case a new file channel is opened.
+ *
+ * @author piegames
+ */
+public class RegionFile implements Iterable<Chunk> {
+
+	protected final Path	file;
+
+	protected final Chunk[]	chunks;
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: Don't cache chunks! getChunk() and setChunk() should directly load and save chunks.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+		timestamps2 = timestamps.asIntBuffer();
+
+		chunks = new Chunk[32 * 32];
+		chunksUpToDate = new boolean[32 * 32];
+		for (int i = 0; i < 1024; i++) {
+			int chunkPos = locations2.get(i) >>> 8;
+			int chunkLength = locations2.get(i) & 0xFF;
+			if (chunkPos > 0) {
+				/* i & 31 retrieves the last 5 bit which store the x coordinate */
+				chunks[i] = new Chunk(i & 31, i >> 5, timestamps2.get(i), raf, chunkPos, chunkLength);
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: See, this is why we don't want to cache chunks. :P
You just loaded the entire region file into memory. And have twice the state to keep track of.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+	 * Returns the {@link Chunk} at the given coordinate
+	 *
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @return the chunk at that coordinate or {@code null} if the chunk does not exist
+	 * @author piegames
+	 */
+	public Chunk getChunk(int x, int z) {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: This method should return Optional<Chunk> instead of a nullable Chunk.
At the vary least annotate this method with `@Nullable`

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+		return chunks[(x & 31) | (z << 5)];
+	}
+
+	public Chunk getChunk(int i) {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: This method should be private. If you want to expose a RegionFile as a fixed length (4096) one dimensional array of blobs and use a higher abstraction wrapper/layer to translate to/from x, y, z (3 dimensions), you could do that, but I don't see any good reason why. If someone reeeeealy wanted to use a region file to store a one dimensional array of blobs they could just as well write there own wrapper to translate to 3 dimensions.

Wait... Why am I going crazy over this? ... Just make this method private! :P

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+	 * 
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @param chunk
+	 *            The chunk to set. Use {@code null} to remove it.
+	 * @author piegames
+	 */
+	public synchronized void setChunk(int x, int z, Chunk chunk) {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: Annotate `chunk` with `@Nullable`. Or even better, add a clearChunk(x, z) method

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+		setChunk((x & 31) | (z << 5), chunk);
+	}
+
+	public synchronized void setChunk(int i, Chunk chunk) {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: Make this method private.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+		chunks[i] = chunk;
+		chunksUpToDate[i] = false;
+	}
+
+	/**
+	 * Write all chunks that got changed to disk, update the file's header (chunk locations and timestamps) and truncate the file at the end
+	 * 
+	 * @author piegames
+	 */
+	public synchronized void writeChanges() throws IOException {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: I would just write chunks directly in setChunk(), and remove this method.
Some of this code can be copy n' pasted into setChunk() but if there is anything we can do to make it a little easer to read that would be nice.

File: src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java [hunk 15]
```diff
@@ -35,6 +35,7 @@
 import com.flowpowered.nbt.Tag;
 import com.flowpowered.nbt.stream.NBTInputStream;
 
+@Deprecated
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java:34]: We can also delete most of the code in this class! :D
Just make it wrap and or use the new one.

File: src/main/java/com/flowpowered/nbt/ByteArrayTag.java [hunk 16]
```diff
@@ -50,6 +50,11 @@ public ByteArrayTag(String name, byte[] value) {
         return value;
     }
 
+    @Override
+    public void setValue(byte[] value) {
```
Comment (xcube16) [inline src/main/java/com/flowpowered/nbt/ByteArrayTag.java:54]: Why did you make every Tag mutable!? D': *cries*
I don't care what @kashike says, this is one of my favorite NBT libraries.
Immutable Tags are one reason for that.

If everyone really wants this.... *cries some more*
Maybe it should be in its own PR?

File: src/main/java/com/flowpowered/nbt/stream/NBTInputStream.java (hunk 4 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/stream/NBTInputStream.java]: There are three ways chunk data can be compressed. And if the `NBTInputStream` provides syntactic sugar to unwrap one compression, I thought that I should add the other one as well (the first time I tried to load region files I didn't know they had custom compression so I thought there was a bug or so, quite painful).
I do agree that this was kind of a poor design decision (the library should not handle compression/endianness at all), but I wanted this change to be backwards-compatible.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 7 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: `usedSectors` is populated with the old state for chunks that haven't changed (see the continue in line 137). If the modified chunks haven't grown, they'll be reinserted there afterwards.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 9 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: > You just loaded the entire region file into memory

In most of my use cases I need to do so anyway so if I do it all at once the OS will optimize it for me. I don't care about RAM usage, honestly.
> And have twice the state to keep track of.

Why? I don't care about the hard disk file once it's loaded.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 10 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: I'd like to, but if I understand it correctly we'd need to add a library to the class path to get those annotations.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 11 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: I originally deleted this method, but then added it back in because I needed it. When iterating over all chunks while needing an index, it's silly to convert that index to a (x,y) pair just to pass it to the method and then convert it back to access the array.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 13 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: Same as above.

File: src/main/java/com/flowpowered/nbt/ByteArrayTag.java (hunk 16 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/ByteArrayTag.java:54]: Have you ever tried to change the position of an entity in chunk with the NBT data being immutable?

> Maybe it should be in its own PR?

IIRC this was, but I merged it onto my dev branch and I later build the API changes on top of it; sorry.

File: src/main/java/com/flowpowered/nbt/regionfile/Chunk.java (hunk 2 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/Chunk.java]: Will do.

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 14 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: I initially thought of this, but if there are frequent changes (or say I modify each chunk a bit), there will be a lot of redundant disk writes and pointless position calculations.

File: src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java (hunk 15 above)
Comment (piegamesde) [reply src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java:34]: Bah, this code is so ugly I do not really want to touch this :D 
It does not even handle missing chunks or exceptions and has no documentation.

**Comment by piegamesde** [comment]
Let me explain the intended workflow with this classes, this might help you understand my design decisions:

- Load the entire file from disk (and parse the header). In most of the cases, I want to access the whole file anyway, so let's make it fast at least. Also, once the file is loaded, I don't care about its state on disk.
- Iterate over all (or some) chunks and process them.
  - Processing starts by converting this to an NBT Tag which gets parsed as needed (this is the probably most expensive part)
  - Analyse or modify it
  - (Optionally) create a new chunk object with the modified data (this will serialize it back to a buffer; expensive)
- (Optionally) write the changed file back to disk.

**Comment by xcube16** [comment]
>     * Load the entire file from disk (and parse the header). In most of the cases, I want to access the whole file anyway, so let's make it fast at least. Also, once the file is loaded, I don't care about its state on disk.

This sounds a lot like what SimpleRegionFileReader.readFile(File) does (reads entire file into a list, all be it with limited metadata for each chunk).

>     * Iterate over all (or some) chunks and process them.
>       
>       * Processing starts by converting this to an NBT Tag which gets parsed as needed (this is the probably most expensive part)
>       * Analyse or modify it
>       * (Optionally) create a new chunk object with the modified data (this will serialize it back to a buffer; expensive)

SimpleRegionFileReader.readFile(File).stream()...

>     * (Optionally) write the changed file back to disk.

All we would need to do is add SimpleRegionFileReader.writeFile(List<Tag<?>>)
(we might want to add an immutable region chunk metadata class to encapsulate Tag<?> providing the key (x, y, z) and timestamp)

This simple approach of just upgrading SimpleRegionFileReader would be easy to do and far less new code. When writing back to disk it would also on average create a more compact region file.

But does that really match what a region file is and what it is used for? From my point of view, a region file is more like a random access mutable key-value database that uses a file for backing.

Think about an NBT editor or another tool that wants to access and modify individual chunks. This is also how Minecraft uses region files. Iteration is not a region file's primary nature by design, but can also be implemented with not much extra code, even with the optimization I suggest below. (should be fun)

There is not much point in having a RegionFile object (as you can see, SimpleRegionFileReader does not) unless you are caching chunks in it (bad! let the user do that! :P), or using it like a database. This is likely where my misunderstanding of your design goals happened.

> let's make it fast at least

Often I find a good design is fast anyway so lets do that last.
But now that I think about it, the fastest way to iterate the chunks and or load them all into memory would be to read them consecutively with regard to there location on disk. Your implementation, and not even SimpleRegionFileReader does that.

File: src/main/java/com/flowpowered/nbt/stream/NBTInputStream.java (hunk 4 above)
Comment (xcube16) [reply src/main/java/com/flowpowered/nbt/stream/NBTInputStream.java]: One could just require the user to instantiate EndianSwitchableInputStream and make NBTInputStream's constructor require a DataInput instead of InputStream. I would still have EndianSwitchableInputStream in the library. The current helper constructors that support endianness internally are not that bad though... If its not broken don't fix it.

You could create a CompressedInputStream class that supports all three compression types. Wait why would you do that? The only methods common to all three types are already in InputStream. See where this is going?

File: src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java (hunk 14 above)
Comment (xcube16) [reply src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java]: Let the user handle that. If there are not frequent changes you are just creating a cache invalidation hell for the user for no reason.
If the user wants to cache chunks, they know the best way to do that.
```
/*not final*/ Chunk cachedChunk = RegionFile#getChunk(x, z);
```
Lets make some changes to to the chunk
```
chunk = makeSomeChanges(chunk);
```
hmmm... another frequent change
```
chunk = anotherFrequentChange(chunk);
```
Now lets save all the changes at once
```
RegionFile#setChunk(x, z, chunk)
```

If you are not convinced yet, think about this...
What happens if you want to cache chunks on the world level instead of at the region level (like what Minecraft does internally)?

Caching the chunks would not be so bad if Chunk was immutable as it would hide the caching form the user. But it would still be a lot of useless (in my opinion) code that would just be there for the off change that someone spams getChunk(). And even if they do, it still might not be a problem (most modern OS's cache hot files anyway).

File: src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java (hunk 15 above)
Comment (xcube16) [reply src/main/java/com/flowpowered/nbt/regionfile/SimpleRegionFileReader.java:34]: All the more reason to get rid of it I would think. An easy one line of code could replace almost that entire file. Doesn't it make make you feel good when you remove lots code?

Maybe I just like seeing code get deleted. :P

As of now this PR adds 1530 new lines of code and only removes 46.

**Comment by piegamesde** [comment]
Once I start loading chunks lazily, things start to get way more complicated. It will at least bring back in the state of the open `FileChannel` but that's okay for me. The implementation will at least need to keep all modified chunks in RAM since there is no way around collecting them into one single write.

**Comment by xcube16** [comment]
> Once I start loading chunks lazily, things start to get way more complicated.

I don't see how it is complicated unless you are caching. Can you give me an example?

> It will at least bring back in the state of the open `FileChannel` but that's okay for me.

I am confused, the only implementation that does not keep an open FileChannel is SimpleRegionFileReader.

> The implementation will at least need to keep all modified chunks in RAM since there is no way around collecting them into one single write.

What? Region files where designed for random access both when reading and writing chunks. Whenever you write a chunk, see if it fits in the old location, and if it is too big, search for a space with enough consecutive empty sectors to put it in.
If you don't want to implement this, replace the setChunk() method with a setChunks() method that requires all chunks upfront (no random access, but at least it would be better than caching).

The only state a region file abstraction should expose to the user is the region file its self. Caching should be in its own abstraction layer.
